src/core/ext/filters/backend_metrics/backend_metric_filter.cc

```diff
@@ -0,0 +1,108 @@
+#include "src/core/ext/filters/backend_metrics/backend_metric_filter.h"
+
+#include <grpc/support/port_platform.h>
+
+#include "upb/upb.h"
+#include "upb/upb.hpp"
+#include "xds/data/orca/v3/orca_load_report.upb.h"
+
+#include "src/core/ext/filters/client_channel/lb_policy/backend_metric_data.h"
+#include "src/core/lib/debug/trace.h"
+#include "src/core/lib/promise/map.h"
+#include "src/core/lib/resource_quota/arena.h"
+
+namespace grpc_core {
+
+TraceFlag grpc_backend_metric_filter_trace(false, "backend_metric_filter");
+
+absl::optional<std::string> BackendMetricFilter::MaybeSerializeBackendMetrics(
+    BackendMetricProvider* provider) const {
+  if (provider == nullptr) return absl::nullopt;
+  BackendMetricData d = provider->GetBackendMetricData();
+  upb::Arena arena;
+  xds_data_orca_v3_OrcaLoadReport* response =
+      xds_data_orca_v3_OrcaLoadReport_new(arena.ptr());
+
+  if (d.cpu_utilization != -1) {
+    xds_data_orca_v3_OrcaLoadReport_set_cpu_utilization(response,
+                                                        d.cpu_utilization);
+  }
+  if (d.mem_utilization != -1) {
+    xds_data_orca_v3_OrcaLoadReport_set_mem_utilization(response,
+                                                        d.mem_utilization);
+  }
+  if (d.qps != -1) {
+    xds_data_orca_v3_OrcaLoadReport_set_rps_fractional(response, d.qps);
+  }
+  for (const auto& p : d.request_cost) {
+    xds_data_orca_v3_OrcaLoadReport_request_cost_set(
+        response,
+        upb_StringView_FromDataAndSize(p.first.data(), p.first.size()),
+        p.second, arena.ptr());
+  }
+  for (const auto& p : d.utilization) {
+    xds_data_orca_v3_OrcaLoadReport_utilization_set(
+        response,
+        upb_StringView_FromDataAndSize(p.first.data(), p.first.size()),
+        p.second, arena.ptr());
+  }
+  size_t len;
+  char* buf =
+      xds_data_orca_v3_OrcaLoadReport_serialize(response, arena.ptr(), &len);
```

Comment:
Given that we are going to be unconditionally creating the `CallMetricRecorder` whenever call metric reporting is enabled, even if the application does not actually provide any metrics, I think we should add an optimization here to avoid serializing the proto (or therefore adding the metadata) if none of the fields were actually populated.

Reply:
Still need to address this comment.

Reply:
Done.